### ISSUE TITLE
PLR6301 doesn't detect @override from TYPE_CHECKING

### ISSUE THREAD

**Issue by ItsDrike** [astral-sh/ruff#24713]
### Summary

The preview rule `PLR6301` is meant to detect when a method should instead become a staticmethod, class method or a fucntion, and triggering when the `self` argument in the function body is unused.

The documentation for this rule explains that it doesn't trigger for methods decorated with `@typing.override`, which is good, however, this detection doesn't work when the `override` symbol is imported within a `TYPE_CHECKING` block, and has a fallback in the else block after to just a no-op decorator.

This pattern is pretty important for libraries that need to handle backwards compatibility and don't want to have `typing_extensions` become a runtime dependency.

E.g.

```python
if TYPE_CHECKING:
    from typing_extensions import override
else:
    override = lambda f: f

# Ruff doesn't seem to understand that `override` in here is actually the `typing.override`
# and when PLR6301 violation is detected, even if decorated with `@override`, it triggers.
```

Playground link: <https://play.ruff.rs/45ead6a4-9511-49be-bf76-010f6d325a80>

### Version

ruff 0.15.9

**Comment by ntBre** [comment]
Thanks for the report!

I don't think this is an issue with `PLR6301` or with our `TYPE_CHECKING` handling, at least directly. Instead, this is caused by the `else` branch of the `if TYPE_CHECKING` block. That seems to be shadowing the first definition in a way that our semantic model can't follow.

If you invert the condition and order, it works as expected:

```py
if not TYPE_CHECKING:
    override = lambda f: f  # noqa: E731
else:
    from typing_extensions import override
```

or if you import `override` from one of the known [typing-modules](https://docs.astral.sh/ruff/settings/#lint_typing-modules), but those are both workarounds.

This may be pretty tricky to fix without introducing other issues, but it does feel like a bug to me.

**Comment by MichaReiser** [comment]
I don't know if this is an option for you but you could have your own `typing` modules that only exportes `override` (and any other typing symbols) from within `TYPE_CHECKING`, then use [`typing_modules`](https://docs.astral.sh/ruff/settings/#lint_typing-modules) to teach ruff that said model should be treated like the built in `typing` module

**Comment by ItsDrike** [comment]
> I don't know if this is an option for you but you could have your own `typing` modules that only exportes `override` (and any other typing symbols) from within `TYPE_CHECKING`, then use [`typing_modules`](https://docs.astral.sh/ruff/settings/#lint_typing-modules) to teach ruff that said model should be treated like the built in `typing` module

I do actually like this idea, I think I'll go with that, though I'd like to keep this open, as it still feels like ruff isn't doing the right thing for what the code is doing.
